### PR TITLE
Add spring boot integration example

### LINKED ISSUE #59: add example of a simple custom Handler

### LINKED ISSUE #82: How to run custom resolvers along with neo4j-graphql-java library?
Is it possible to have my custom resolvers (using graphql-java) integrated and work in parallel with neo4j-graphql-java library? I want to use neo4j-graphql-java library for resolving queries for all objects, however for few of them i want to write my own resolvers.   

The [FAQ on github](https://github.com/neo4j-graphql/neo4j-graphql-java/#how-does-this-related-to-graphql-java) section says "If you wanted, you could combine graphql-java resolvers with this library to have a part of your schema handled by Neo4j"

From this statement i am assuming it can be done. If it is possible, please help me out on how can we do it?

### LINKED ISSUE #97: integration in JAVA
Are there any examples of integration in JAVA? I have encountered some problems in integration and would like to refer to them.I want to see the parsing processes.

How to use spring-boot integration .

### LINKED ISSUE #119: problems using DataFetcherInterceptorDemo
I tried using something based on DataFetcherInterceptorDemo and it didn't seem to work for me.

One difference is that I am using org.neo4j.ogm.session.Session because that is what org.springframework.boot:spring-boot-starter-data-neo4j easily exposes. But it still seems like it should work. Here is my impl:

```
import graphql.language.VariableReference
import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLNonNull
import graphql.schema.GraphQLType
import org.neo4j.graphql.Cypher
import org.neo4j.graphql.DataFetchingInterceptor
import org.neo4j.ogm.session.SessionFactory
import java.math.BigDecimal
import java.math.BigInteger

class CypherDataFetchingInterceptor(val sessionFactory: SessionFactory) : DataFetchingInterceptor {

    override fun fetchData(env: DataFetchingEnvironment, delegate: DataFetcher<Cypher>): Any? {
        val cypher = delegate.get(env)
        val session = sessionFactory.openSession()
        val result = session.query(cypher.query, cypher.params.mapValues { toBoltValue(it.value, env.variables) })
        val data = result.queryResults().map { it.entries }.flatten()
        return if (isListType(cypher.type)) {
            data
        } else {
            data.firstOrNull() ?: emptyMap<String, Any>()
        }
    }

    fun toBoltValue(value: Any?, params: Map<String, Any?>) = when (value) {
        is VariableReference -> params[value.name]
        is BigInteger -> value.longValueExact()
        is BigDecimal -> value.toDouble()
        else -> value
    }

    private fun isListType(type: GraphQLType?): Boolean {
        return when (type) {
            is GraphQLType -> true
            is GraphQLNonNull -> isListType(type.wrappedType)
            else -> false
        }
    }
}
```

This actually works in itself, though I am not sure if I am returning the right kind of Any in both the list and nonlist cases. The problem is that it continues through to graphql-java's AsyncExecutionStrategy.completeValue() which then tries to gather subfields of the result and ends up returning null.

The way I ultimately made querying work was implementing a subclass of AsyncExecutionStrategy and basically skipping further evaluation once the Cypher is executed.

```
import graphql.ExecutionResultImpl
import graphql.execution.*
import graphql.execution.FieldValueInfo.CompleteValueType
import io.armory.astrolabe.neo4j.Neo4jService
import org.neo4j.graphql.Cypher
import java.util.concurrent.CompletableFuture

class CypherExecutionStrategy(val neo4jService: Neo4jService) : AsyncExecutionStrategy() {

    override fun completeValue(executionContext: ExecutionContext, parameters: ExecutionStrategyParameters): FieldValueInfo {
        val result = executionContext.valueUnboxer.unbox(parameters.source)
        return if (result is Cypher) {
            completeValueForCypher(executionContext, parameters, result)
        }
        else {
            super.completeValue(executionContext, parameters)
        }
    }

    fun completeValueForCypher(executionContext: ExecutionContext?, parameters: ExecutionStrategyParameters, cypher: Cypher): FieldValueInfo {
        val fieldValueInfoBuilder = FieldValueInfo.newFieldValueInfo(CompleteValueType.LIST)
        val neo4jResult = neo4jService.query(cypher)
        val data = neo4jResult.queryResults().map { it.values }.flatten()
        try {
            parameters.nonNullFieldValidator.validate(parameters.path, data)
        } catch (e: NonNullableFieldWasNullException) {
            return fieldValueInfoBuilder.fieldValue(Async.exceptionallyCompletedFuture(e)).build()
        }
        return fieldValueInfoBuilder
                .fieldValue(CompletableFuture.completedFuture(ExecutionResultImpl(data, null)))
                .build()
    }
}
```
```
import org.neo4j.graphql.Cypher
import org.neo4j.ogm.model.Result
import org.neo4j.ogm.session.SessionFactory
import org.springframework.stereotype.Component

@Component
class Neo4jService(val sessionFactory: SessionFactory) {
    val session = sessionFactory.openSession()

    fun query(cypher: Cypher): Result {
        return session.query(cypher.query, cypher.params)
    }
}
```

Is there a better way for me to handle this that is more inline with your documentation?

### PR DESCRIPTION
**Pull request by Andy2003**
Finally here is an integration example for spring boot.

* resolves #119 
* resolves #97 
* resolves #59
* resolves #82
* With some adjustments you can use this example with spring data (#66)

### CODE REVIEW

**Comment by Andy2003** [comment]
@claymccoy you may review this PR

**Comment by Andy2003** [comment]
resolves #59
resolves #82

File: examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
+package org.neo4j.graphql.examples.graphqlspringboot.config
+
+import com.expediagroup.graphql.SchemaGeneratorConfig
```
Comment (jexp) [inline examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt:3]: is that a common library used? and recommended by the graphql-java folks?

File: examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+@Configuration
+open class GraphQLConfiguration {
+    private val logger = LoggerFactory.getLogger(GraphQLConfiguration::class.java)
+
+    /**
+     * This generates the augmented neo4j schema
+     */
+    @Bean
+    open fun neo4jSchema(
+            @Value("classpath:schema.graphqls") graphQl: Resource,
```
Comment (jexp) [inline examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt:38]: can we use schema.graphql ?

File: examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/Neo4jConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import java.math.BigInteger
+
+/**
+ * Configuration of the neo4j driver and the DataFetchingInterceptor
+ */
+@Configuration
+@EnableConfigurationProperties(Neo4jProperties::class)
+open class Neo4jConfiguration {
+
+    @ConditionalOnMissingBean
```
Comment (jexp) [inline examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/Neo4jConfiguration.kt:25]: can we use the spring boot starter for neo4j driver instead?

File: examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/Neo4jProperties.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.neo4j.graphql.examples.graphqlspringboot.config
```
Comment (jexp) [inline examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/Neo4jProperties.kt:1]: also from the existing starter

File: examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                    "data": {
+                        "team": [ { "id": "team1" } ]
+                    }
+                }
+                """.trimIndent())
+    }
+
+    private fun assertQuery(query: String, expectedResult: String) {
+        val request = GraphQLRequest(query = query)
+        testClient.post()
```
Comment (jexp) [inline examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt:93]: would it make sense to use the graphql client libray as a test client?

File: examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+import org.springframework.test.web.reactive.server.WebTestClient
+import org.testcontainers.containers.Neo4jContainer
+import org.testcontainers.junit.jupiter.Container
+import org.testcontainers.junit.jupiter.Testcontainers
+import java.net.URI
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@EnableAutoConfiguration
+@Testcontainers
+internal class QueriesTest(
```
Comment (jexp) [inline examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt:23]: can we indicate that this is a integration test?

**Review by jexp — APPROVED** [review]
Looks good, can you have a look at my comments?

File: examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt (hunk 1 above)
Comment (Andy2003) [reply examples/graphql-spring-boot/src/main/kotlin/org/neo4j/graphql/examples/graphqlspringboot/config/GraphQLConfiguration.kt:3]: This lib has many stars [(821)](https://github.com/ExpediaGroup/graphql-kotlin) and simplifies the generation of custom resolver methods.

With the [lib "recommented" by the graphql-java folks](https://www.graphql-java.com/tutorials/getting-started-with-spring-boot/), the wiring of the custom resolvers would have to be done by hand, which can become time-consuming and annoying.

How would you determine if the library is recommended?

File: examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt (hunk 5 above)
Comment (Andy2003) [reply examples/graphql-spring-boot/src/test/kotlin/org/neo4j/graphql/examples/graphqlspringboot/controller/QueriesTest.kt:93]: I don't think we need an additional lib just to trigger the endpoint.

**Pull request merged**